concurrent/src/test/java/io/airlift/concurrent/TestLocks.java

```diff
@@ -0,0 +1,168 @@
+        String stringResult = locking(lock, () -> "foo");
+
+        try {
+            lockingInterruptibly(lock, () -> {
+            });
+        }
+        catch (InterruptedException ignore) {
+        }
+
+        try {
+            lockingInterruptibly(lock, () -> "foo");
+        }
+        catch (InterruptedException ignore) {
+        }
+    }
+
+    @Test
+    public void testLockingWithReturnedValue()
+    {
+        assertEquals(locking(new ReentrantLock(), () -> "foo"), "foo");
+    }
+
+    @Test(expectedExceptions = RuntimeException.class, expectedExceptionsMessageRegExp = "exception thrown under a lock")
+    public void testLockingWithRuntimeException()
+    {
+        locking(new ReentrantLock(), () -> {
+            throw new RuntimeException("exception thrown under a lock");
+        });
+    }
+
+    @Test(expectedExceptions = IOException.class, expectedExceptionsMessageRegExp = "exception thrown under a lock")
+    public void testLockingWithCheckedException()
+            throws IOException
+    {
+        locking(new ReentrantLock(), () -> {
+            throw new IOException("exception thrown under a lock");
+        });
+    }
+
+    @Test
+    public void testLockingLocks()
+            throws Exception
+    {
+        Lock lock = new ReentrantLock();
+        int threads = 10;
+        CyclicBarrier barrier = new CyclicBarrier(threads);
+        AtomicInteger underLock = new AtomicInteger(0);
+        AtomicInteger maxUnderLock = new AtomicInteger(0);
+
+        List<Future<?>> futures = IntStream.range(0, threads)
+                .mapToObj(i -> executor.submit(() -> {
+                    barrier.await();
+                    locking(lock, () -> {
+                        maxUnderLock.accumulateAndGet(underLock.incrementAndGet(), Integer::max);
+                        Thread.sleep(30);
+                        underLock.decrementAndGet();
+                    });
+
+                    return null;
+                }))
+                .collect(toImmutableList());
+
+        futures.forEach(MoreFutures::getFutureValue);
+    }
+
+    @Test
+    public void testLockingInterruptiblyCanBeInterrupted()
+            throws Exception
+    {
+        Lock lock = new ReentrantLock();
+        Exchanger<Thread> testBackgroundThreadExchanger = new Exchanger<>();
+
+        Future<Object> future = executor.submit(() -> {
+            testBackgroundThreadExchanger.exchange(Thread.currentThread());
+            // main thread is under lock now
+            lockingInterruptibly(lock, () -> {
+                throw new AssertionError("test background thread got the lock");
+            });
+            throw new AssertionError("lockingInterruptibly returned, InterruptedException was expected");
+        });
+
+        locking(lock, () -> {
+            Thread testBackgroundThread = testBackgroundThreadExchanger.exchange(null, 5, SECONDS);
+            Thread.sleep(40); // give the thread chance to lock
```

Comment:
I agree. However, in this case I expect it to work even if we interrupt to early. The reason is that `Lock#lockInterruptibly` throws `InterruptedException` even if the lock is available.
The sleep is here to make it "almost certain" thread is already waiting for a lock so that we test the (expectedly) most common case. However, test passes even with sleep removed.